libs/entitlement/tools/securitymanager-scanner/src/main/java/org/elasticsearch/entitlement/tools/securitymanager/scanner/SecurityCheckClassVisitor.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.entitlement.tools.securitymanager.scanner;
+
+import org.elasticsearch.core.SuppressForbidden;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Type;
+
+import java.lang.constant.ClassDesc;
+import java.lang.reflect.InaccessibleObjectException;
+import java.lang.reflect.Modifier;
+import java.nio.file.Path;
+import java.security.Permission;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.objectweb.asm.Opcodes.ACC_PUBLIC;
+import static org.objectweb.asm.Opcodes.ASM9;
+import static org.objectweb.asm.Opcodes.GETSTATIC;
+import static org.objectweb.asm.Opcodes.INVOKEDYNAMIC;
+import static org.objectweb.asm.Opcodes.INVOKEINTERFACE;
+import static org.objectweb.asm.Opcodes.INVOKESPECIAL;
+import static org.objectweb.asm.Opcodes.INVOKESTATIC;
+import static org.objectweb.asm.Opcodes.INVOKEVIRTUAL;
+import static org.objectweb.asm.Opcodes.NEW;
+
+class SecurityCheckClassVisitor extends ClassVisitor {
+
+    static final String SECURITY_MANAGER_INTERNAL_NAME = "java/lang/SecurityManager";
+    static final Set<String> excludedClasses = Set.of(SECURITY_MANAGER_INTERNAL_NAME);
+
+    record CallerInfo(
+        String moduleName,
+        String source,
+        int line,
+        String className,
+        String methodName,
+        boolean isPublic,
+        String permissionType,
+        String runtimePermissionType
+    ) {}
+
+    private final Map<String, List<CallerInfo>> callerInfoByMethod;
+    private String className;
+    private int classAccess;
+    private String source;
+    private String moduleName;
+    private String sourcePath;
+    private Set<String> moduleExports;
+
+    protected SecurityCheckClassVisitor(Map<String, List<CallerInfo>> callerInfoByMethod) {
+        super(ASM9);
+        this.callerInfoByMethod = callerInfoByMethod;
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+        super.visit(version, access, name, signature, superName, interfaces);
+        this.className = name;
+        this.classAccess = access;
+    }
+
+    @Override
+    public void visitSource(String source, String debug) {
+        super.visitSource(source, debug);
+        this.source = source;
```

Comment:
I think the typical style for this sort of thing is to set these kinds of context variables before calling the `super` method. But I suspect it doesn't matter in this case.